### ISSUE TITLE
Debian buster/sid Bug

### ISSUE THREAD

**Issue by sahal8020** [mhogomchungu/zuluCrypt#110]
I'm facing the exact issue reported here:
https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=922038

Right now the latest version of zuluCrypt is completely inoperable. I downgraded to debian stable version which made zuluCrypt work for opening luks volumes but trying to open zuluCrypt created plain dm-crypt I get `Insufficient privilege to open a system volume.`

Is there any workaround in the mean time while this is being (hopefully) resolved?

Thank you.

**Comment by mhogomchungu** [comment]
That symbol belongs to libgcrypt.

Was libgcrypt recently updated in your system?

The easiest solution is to recompille zuluCrypt against the updated version of libcrypt.

**Comment by sahal8020** [comment]
The upgrade after which all hell broke loose:

```
cryptsetup-bin:amd64 (2:2.0.6-1, 2:2.1.0-1),
cryptsetup-initramfs:amd64 (2:2.0.6-1, 2:2.1.0-1),
libgcc-7-dev:amd64 (7.4.0-4, 7.4.0-5),
exim4-base:amd64 (4.92~RC5-2, 4.92~RC6-1),
libc6-dev:amd64 (2.28-6, 2.28-7),
cpp-7:amd64 (7.4.0-4, 7.4.0-5),
libc6:amd64 (2.28-6, 2.28-7),
gcc-7-base:amd64 (7.4.0-4, 7.4.0-5),
locales:amd64 (2.28-6, 2.28-7),
cryptsetup-run:amd64 (2:2.0.6-1, 2:2.1.0-1),
libc-bin:amd64 (2.28-6, 2.28-7),
gcc-7:amd64 (7.4.0-4, 7.4.0-5),
libc-dev-bin:amd64 (2.28-6, 2.28-7),
multiarch-support:amd64 (2.28-6, 2.28-7),
libcryptsetup12:amd64 (2:2.0.6-1, 2:2.1.0-1),
libdebconfclient0:amd64 (0.246, 0.247),
...
```
Sadly I'm unable to revert it as many of those files no longer exist online. 

As for libgcrypt, the one I have is libgcrypt20 (1.8.4-5) which has been out for a while according to its [log](https://metadata.ftp-master.debian.org/changelogs//main/libg/libgcrypt20/libgcrypt20_1.8.4-5_changelog). I update regularly so not sure if it's the main culprit.

**Comment by mhogomchungu** [comment]
zuluCrypt has two direct dependencies:
1. libcryptsetup
2. libgcrypt

zuluCrypt has two optional direct dependencies:
1. kwallet
2. libsecret

The problem started after you update cryptsetup among other packages and i suspect the source of the problem is here. 

A solution is like i said, rebuild zuluCrypt. Problems were also reported in arch linux after a system update[1] and rebuilding zuluCrypt has worked there.

[1] https://aur.archlinux.org/packages/zulucrypt/

**Comment by sahal8020** [comment]
Rebuilding did resolve the problem :+1: .

Not sure if I should close this as probably others on debian are facing the same thing and might not know how to resolve it. Might be a good idea to rebuild the packages on your end and push them to the appropriate repos. 

I'll leave the decision to you. I'm good (for now :stuck_out_tongue_closed_eyes:).

**Comment by Ahzod** [comment]
Even though this was triggered by an upgrade of libcryptsetup, the actual problem is within zuluCrypt.

Cryptsetup switched from using libgcrypt to libssl and thus no longer links against the former. However, libzuluCrypt still uses libgcrypt, but the debian version does not link directly against libgcrypt. This is because the debian package uses zuluplay-static and is built with the '-Wl,--as-needed' linker flag, which exposes an ordering problem in the link target for libzuluCrypt: '-lgcrypt' and the other shared libraries are listed before 'zuluplay-static' in the TARGET_LINK_LIBRARIES for zuluCrypt, thus the linker doesn't consider the functions used in zuluplay-static, when determining that libgcrypt is not used and thus need not be linked against.

The solution is to simply move zuluplay-static to the other static libraries (String StringList Process) and before the shared ones.

**Comment by JackDinn** [comment]
Just wanted to say thx for this thread, I don't know if its even related as i never use Zulucrypt via console but i was getting this error with the GUI
![zz_20190303-183841](https://user-images.githubusercontent.com/5557560/53699854-779b1200-3de4-11e9-81b4-7d82fed63457.png)

I re-built from the AUR repo https://aur.archlinux.org/packages/zulucrypt-git and it's working again, Pheww I thought i had lost access to all my vault files.

using, Manjaro KDE

**Comment by peekaa** [comment]
I have the same issue here, in debian 10. How to rebuild? I have never done it before... :-( Is there any chance to fix it in a week or two in debian? Its really problem, i need files is tc files....

**Comment by mhogomchungu** [comment]
There is already a bug report open at debian's bug reported and hence the best way forward is to ask and plead with the maintainer to rebuild the project.

If you want to rebuild from source, the steps are pretty simple and they are outlined here[1]

[1] https://github.com/mhogomchungu/zuluCrypt/blob/master/BUILD_INSTRUCTIONS

**Comment by Ahzod** [comment]
Merely rebuilding zulucrypt doesn't fix the problem seen in debian. One first needs to fix zuluCrypt-cli/CMakeLists.txt, see pull request #112.

Meanwhile, it is possible to work around this bug by preloading libgcrypt:
# LD_PRELOAD=/lib/x86_64-linux-gnu/libgcrypt.so.20 zuluCrypt-cli

**Comment by mhogomchungu** [comment]
The issue was automatically closed when i accepted the pull request.

Reopening it and will keep it open until debian upstream pushes an update with a fix.

**Comment by peekaa** [comment]
No, not working here in Debian 10. Installed from official repo,last update today.

**Comment by mhogomchungu** [comment]
A bit of googling led me here[1]. Its in "testing" and will show up in about 4 days.

[1]https://tracker.debian.org/pkg/zulucrypt

**Comment by Ahzod** [comment]
The fixed zuluCrypt version 5.4.0-3 has migrated from Debian unstable to Debian testing, so this bug can now be closed.
Thanks for all your work on zuluCrypt!
